exonum/src/storage/proof_map_index/key.rs

```diff
@@ -76,53 +67,109 @@ impl ::std::ops::Not for ChildKind {
     }
 }
 
+/// A structure that represents paths to the any kinds of `ProofMapIndex` nodes.
+#[derive(Copy, Clone)]
+pub struct DBKey {
+    bytes: [u8; DB_KEY_SIZE],
+    start: u16,
+}
+
 impl DBKey {
-    /// Create a new bit slice from the given binary data.
-    pub fn leaf<K: ProofMapKey>(key: &K) -> DBKey {
+    /// Create a path from the given key.
+    pub fn new<K: ProofMapKey>(key: &K) -> DBKey {
         debug_assert_eq!(key.size(), KEY_SIZE);
 
-        let mut data = [0; KEY_SIZE];
-        key.write(&mut data);
+        let mut data = [0; DB_KEY_SIZE];
+        data[0] = LEAF_KEY_PREFIX;
+        key.write(&mut data[1..KEY_SIZE + 1]);
+        data[KEY_SIZE + 1] = 0;
+        DBKey::from_raw(data)
+    }
+
+    /// Shows the type of path.
+    pub fn is_leaf(&self) -> bool {
+        self.bytes[0] == LEAF_KEY_PREFIX
+    }
+
+    /// Returns the byte representation of `DBKey`.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.bytes
+    }
+
+    /// Constructs the `DBKey` from raw bytes.
+    fn from_raw(raw: [u8; DB_KEY_SIZE]) -> DBKey {
         DBKey {
-            data: data,
-            from: 0,
-            to: (KEY_SIZE * 8) as u16,
+            bytes: raw,
+            start: 0,
         }
     }
 
-    #[doc(hidden)]
-    pub fn from(&self) -> u16 {
-        self.from
+    /// Sets the right border of the bit range.
+    fn set_end(&mut self, end: u16) {
+        let max_len = (KEY_SIZE * 8) as u16;
+        assert!(end <= max_len);
+        // Update DBKey kind and right bound.
+        if end == max_len {
+            self.bytes[DB_KEY_SIZE - 1] = 0;
+            self.bytes[0] = LEAF_KEY_PREFIX;
+        } else {
+            self.bytes[0] = BRANCH_KEY_PREFIX;
+            self.bytes[DB_KEY_SIZE - 1] = end as u8;
+        };
     }
+}
 
-    // TODO: terrible hack, try to remove this (ECR-22)
-    #[doc(hidden)]
-    pub fn set_from(&mut self, from: u16) {
-        self.from = from
+/// The bits representation of the `DBKey`.
+pub trait KeyBitRange {
```

Comment:
Is this trait intended to be `pub` (available to service implementers)? Perhaps it makes sense to make it `pub(crate)`?